src/ethereum/prague/vm/eof.py

```diff
@@ -0,0 +1,580 @@
+    type_section_contents = []
+    for _ in range(type_size // 4):
+        type_section_contents.append(container[counter : counter + 4])
+        counter += 4
+
+    if len(container) < counter + sum(code_sizes):
+        raise InvalidEOF("Code section size does not match header")
+    code_section_contents = []
+    for code_size in code_sizes:
+        code_section_contents.append(container[counter : counter + code_size])
+        counter += code_size
+
+    if container_sizes:
+        if len(container) < counter + sum(container_sizes):
+            raise InvalidEOF("Container section size does not match header")
+        container_section_contents = []
+        for container_size in container_sizes:
+            container_section_contents.append(
+                container[counter : counter + container_size]
+            )
+            counter += container_size
+    else:
+        container_section_contents = None
+
+    if len(container) < counter + data_size:
+        raise InvalidEOF("Data section size does not match header")
+    data_section_contents = container[counter : counter + data_size]
+    counter += data_size
+
+    # Check for stray bytes after the data section
+    if len(container) > counter:
+        raise InvalidEOF("Stray bytes found after data section")
+
+    return EOFMetadata(
+        type_size=type_size,
+        num_code_sections=num_code_sections,
+        code_sizes=code_sizes,
+        num_container_sections=num_container_sections,
+        container_sizes=container_sizes,
+        data_size=data_size,
+        body_start_index=body_start_index,
+        type_section_contents=type_section_contents,
+        code_section_contents=code_section_contents,
+        container_section_contents=container_section_contents,
+        data_section_contents=data_section_contents,
+    )
+
+
+def validate_body(eof_meta: EOFMetadata) -> None:
+    """
+    Validate the body of the EOF container.
+
+    Parameters
+    ----------
+    eof_meta : EOFMetadata
+        The metadata of the EOF container.
+
+    Raises
+    ------
+    InvalidEOF
+        If the EOF container is invalid.
+    """
+    for i, type_section in enumerate(eof_meta.type_section_contents):
+        num_inputs = type_section[0]
+        if num_inputs > 127:
+            raise InvalidEOF(f"Invalid number of inputs for type {i}")
+
+        num_outputs = type_section[1]
+        if num_outputs > 128:
+            raise InvalidEOF(f"Invalid number of outputs for type {i}")
+
+        max_stack_height = Uint.from_be_bytes(type_section[2:])
+        if max_stack_height > 1023:
+            raise InvalidEOF(f"Invalid max stack height for type {i}")
+
+        # 4750: Input and output for first section should
+        # be 0 and 128 respectively
+        if i == 0 and (num_inputs != 0 or num_outputs != 0x80):
+            raise InvalidEOF("Invalid input/output for first section")
+
+
+def get_valid_instructions(code: bytes) -> Set[int]:
```

Comment:
Should have a better name for this. From just the signature, it looks like it returns a uniqued list of all the opcodes used in `code`, but I think it returns offsets instead?

Reply:
Has now been integrated into the `validate_code_section` function.